src/app/clusters/test-cluster-server/test-cluster-server.cpp

```diff
@@ -334,17 +340,58 @@ CHIP_ERROR TestAttrAccess::WriteListStructOctetStringAttribute(AttributeValueDec
 CHIP_ERROR TestAttrAccess::ReadListNullablesAndOptionalsStructAttribute(AttributeValueEncoder & aEncoder)
 {
     return aEncoder.EncodeList([](const auto & encoder) -> CHIP_ERROR {
-        // Just encode a single default-initialized
-        // entry for now.
-        Structs::NullablesAndOptionalsStruct::Type entry;
-        ReturnErrorOnFailure(encoder.Encode(entry));
+        // Just encode our one struct for now.
+        ReturnErrorOnFailure(encoder.Encode(gNullablesAndOptionalsStruct));
         return CHIP_NO_ERROR;
     });
 }
 
 CHIP_ERROR TestAttrAccess::WriteListNullablesAndOptionalsStructAttribute(AttributeValueDecoder & aDecoder)
 {
-    // TODO Add yaml test case for NullablesAndOptionalsStruct list
+    DataModel::DecodableList<Structs::NullablesAndOptionalsStruct::DecodableType> list;
+    ReturnErrorOnFailure(aDecoder.Decode(list));
+
+    size_t count;
+    ReturnErrorOnFailure(list.ComputeSize(&count));
+    // This should really send proper errors on invalid input!
+    VerifyOrReturnError(count == 1, CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto iter = list.begin();
+    while (iter.Next())
+    {
+        auto & value = iter.GetValue();
+        // We only support some values so far.
+        VerifyOrReturnError(value.nullableString.IsNull(), CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrReturnError(!value.optionalString.HasValue(), CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrReturnError(!value.nullableOptionalString.HasValue(), CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrReturnError(value.nullableStruct.IsNull(), CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrReturnError(!value.optionalStruct.HasValue(), CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrReturnError(!value.nullableOptionalStruct.HasValue(), CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrReturnError(!value.optionalList.HasValue(), CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrReturnError(!value.nullableOptionalList.HasValue(), CHIP_ERROR_INVALID_ARGUMENT);
+
+        // Start our value off as null, just in case we fail to decode things.
+        gNullablesAndOptionalsStruct.nullableList.SetNull();
+        if (!value.nullableList.IsNull())
+        {
+            ReturnErrorOnFailure(value.nullableList.Value().ComputeSize(&count));
+            VerifyOrReturnError(count <= ArraySize(gSimpleEnums), CHIP_ERROR_INVALID_ARGUMENT);
+            auto iter2       = value.nullableList.Value().begin();
+            gSimpleEnumCount = 0;
+            while (iter2.Next())
+            {
+                gSimpleEnums[gSimpleEnumCount] = iter2.GetValue();
+                ++gSimpleEnumCount;
+            }
+            ReturnErrorOnFailure(iter2.GetStatus());
+            gNullablesAndOptionalsStruct.nullableList.SetNonNull(Span<SimpleEnum>(gSimpleEnums, gSimpleEnumCount));
```

Comment:
`gSimpleEnumCount` ends up beeing equals to `ArraySize(gSimpleEnums)` ? Can't we just use a local variable to use it as an index for pushing things into `gSimpleEnums` or even `value.nullableList.Value().begin() - iter2` instead of a global variable ? I may have missed why it needs to be global though.

Reply:
> `gSimpleEnumCount` ends up beeing equals to `ArraySize(gSimpleEnums)` ?

No, it ends up some value that is at most the size of that array.  That is, we allow a consumer to write a list that is shorter than the max that gSimpleEnums can fit, and we need to keep track of how many items are actually set in gSimpleEnums when the value is later read.